Block/BlockBase/Block receive messages

shargon opened the pull request:
- Block and BlockHeader refactorization.
- Unmap ipv6 address when don't needed (fail on my computer)

![image](https://user-images.githubusercontent.com/3167973/42555027-3e8148f6-84e7-11e8-840e-4908dce8e71a.png)

shargon committed:
Merge pull request #10 from CityOfZion/development

SYNC

shargon committed:
Merge pull request #12 from CityOfZion/development

Sync

shargon committed:
Merge pull request #13 from CityOfZion/development

SYNC

shargon committed:
Merge pull request #14 from CityOfZion/development

SYNC

shargon committed:
Genesis

shargon committed:
Merge branch 'development' into development

shargon committed:
Add genesis to blockchain

shargon committed:
Merge pull request #15 from CityOfZion/development

Genesis on blockchain (#149)

shargon committed:
Merge pull request #16 from CityOfZion/development

Fix server (#147)

shargon committed:
Merge pull request #17 from CityOfZion/development

SYNC

shargon committed:
Merge pull request #18 from CityOfZion/development

Fixing RedisDbRepository need on NeoSharpRepository (#161)

shargon committed:
Merge pull request #19 from CityOfZion/development

SYNC

shargon committed:
Merge pull request #20 from CityOfZion/development

Sync

shargon committed:
Merge pull request #22 from CityOfZion/development

Sync

shargon committed:
Merge pull request #23 from CityOfZion/development

Make RocksDbRepository testable. (#173)

shargon committed:
fix-autocomplete

shargon committed:
Merge pull request #25 from CityOfZion/development

Sync

shargon committed:
Merge pull request #26 from CityOfZion/development

Fix json print (#181)

shargon committed:
Console Fix

- Is not working well on unix and MacOS
- Console.KeyAvailable is not working as expected

shargon committed:
Merge pull request #27 from CityOfZion/development

sync

shargon committed:
Merge pull request #28 from CityOfZion/development

All IRepository calls use the pattern Async/Await (#182)

shargon committed:
Fixes BlockchainBase / GetPeers without Ipv6

- BlockHeaderBase removed from Blockchain
- Unmap ipv6 address when don't needed (fail on my computer)

shargon committed:
clean code

shargon committed:
fix

shargon committed:
Clean code

osmirnov commented on src/NeoSharp.Core/Messaging/Handlers/BlockHeadersMessageHandler.cs:
```diff
@@ -11,23 +13,51 @@ namespace NeoSharp.Core.Messaging.Handlers
         public async Task Handle(BlockHeadersMessage message, IPeer sender)
         {
-            _blockchain.AddBlockHeaders(message.Payload.Headers.Select(u => u.Header));
+            // Why we need to store BlockHeaderBase? don't have TX hashes, is better ask for blocks and store when we receive it
```
Ok, I looked into NEO project to understand why this works like this:
[https://github.com/neo-project/neo/blob/9cf0d94f0979ea579e6b9d660359e0c49625532b/neo/Network/RemoteNode.cs#L227](https://github.com/neo-project/neo/blob/9cf0d94f0979ea579e6b9d660359e0c49625532b/neo/Network/RemoteNode.cs#L227)

## How we sync block headers

**TL;DR**: VerAckMessageHandler kicks of the cycle of getting block headers and BlockHeadersMessageHandler continues it until gets them all.

GetBlockHeadersMessageHandler/BlockHeadersMessageHandler work in conjunction with VerAckMessageHandler. If the current peer has fewer block headers than the remote one it requests missing block headers sending GetBlockHeadersMessage in VerAckMessageHandler. We expect to get the result in BlockHeadersMessage and to handle in BlockHeadersMessageHandler. BUT, GetBlockHeadersMessageHandler has limitation on the number block headers in response (MaxBlockHeadersCountToReturn = 2000). I suspect it is introduced as the pagination mechanism. This is the reason why we have the duplicate condition in VerAckMessageHandler and BlockHeadersMessageHandler.

## Why we don't care about transaction hashes in block headers

**TL;DR**:  transaction hashes are not critical for blockchain sync.

We have two separate messages for blocks and block headers. Consequently, they are synchronized differently and have different applications. The goal for block headers to sync as fast as possible to get the latest block hash to process other blocks based on it. So on start, the neo client is very busy to get block headers in sync. As opposed to that process, the goal for block sync is to get full fledge blockchain snapshot including transaction information. Needless to say, the process is heavy and is paused when we need to handle new block, transactions, consensus. The block sync operates just a block hash range for fetching. As you see this approach does not rely on transaction hashes. Maybe, we can use them from block headers for different purposes but right now I am not aware of these scenarios.

## Where the block sync is

**TL;DR**:  Not exists for now but we put this in PeerMessageListener

We are going to implement block sync as it was before. Please check neo project for reference
[https://github.com/neo-project/neo/blob/master/neo/Network/RemoteNode.cs#L456](https://github.com/neo-project/neo/blob/master/neo/Network/RemoteNode.cs#L456).
Our place is PeerMessageListener (in while loop). Later it will be optimized not to delay message processing.

**This change will break block header sync logic and every new peer will be overflooded with getting blockchain snapshot until it is able to do any useful workload. So please don't.**

shargon replied:
So i will made some changes, according to your comment

shargon replied:
Now i understand the dummy values :P

shargon replied:
@osmirnov please take a look again, i think that we can change Block like this

```
block
{
 BlockHeader Header{get;set;}
 Transaction[] Tx{get;set;}
}
```

Something like that ..

shargon committed:
Changes

shargon committed:
Block Serialization

- Change Binary serialization (ReadOnly with fields), release stream, enum serializer
- Refactor Block/BlockHeader

shargon committed:
Fix message

shargon committed:
Fix

shargon committed:
Fix contract

shargon committed:
Fix

shargon committed:
Fix BlockHeaderMessage

shargon committed:
ContainsBlock only for blocks

shargon committed:
Message

shargon committed:
Rename Dummy property

shargon committed:
Merge branch 'development' into blockbase

shargon committed:
Merge branch 'development' into blockbase

shargon committed:
Clean code

shargon committed:
Merge branch 'blockbase' of https://github.com/shargon/neo-sharp into blockbase

shargon committed:
Get/Set BlockHeaderHeight

osmirnov commented on src/NeoSharp.Core/Messaging/Handlers/BlockHeadersMessageHandler.cs:
```diff
@@ -1,33 +1,84 @@
+
+            // Already sent on VerAck
+
+            //if (_blockchain.LastBlockHeader.Index < sender.Version.CurrentBlockIndex)
```
Once again, without this condition, only the first 2000 block headers/blocks will be synced.

shargon committed:
Receive Blocks

shargon committed:
BlockMessage

- Only have one block
- Log error on messages
- Receive blocks!

shargon committed:
Clean prompt

shargon committed:
Clean code
